src/shared/sol-iio.h

```diff
@@ -0,0 +1,194 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief Routines to access Linux Industrial I/O (iio) devices under Solleta
+ */
+
+struct sol_iio_device;
+struct sol_iio_channel;
+struct sol_iio_buffer;
+
+/**
+ * Callback to read data from buffer
+ */
+typedef void (*sol_iio_reader_cb)(struct sol_iio_device *device, struct sol_iio_buffer *buffer, void *data);
+
+/**
+ * Open an IIO device
+ *
+ * @param device_id Id of iio device. It's the number which identifies device
+ * on file system. Can be found at '/sys/bus/iio/devices/iio:deviceX'
+ * @param buffer_size size of reading buffer. If 0, will use device default;
+ * if -1 will disable buffer - and all reads will be performed on channel files
+ * on sysfs.
+ * @param trigger_name Name of IIO trigger that will be used on this device. If NULL
+ * or empty, will try to use device current trigger. If device has no current trigger,
+ * will create a 'sysfs_trigger' and use it.
+ *
+ * @return A new IIO handle
+ */
+struct sol_iio_device *sol_iio_open(int device_id, int buffer_size, const char *trigger_name);
+
+/**
+ * Close an IIO device
+ *
+ * @param device IIO device handle to close
+ */
+void sol_iio_close(struct sol_iio_device *device);
+
+/**
+ * Add reading channel.
+ *
+ * @param device IIO handle of device to which channel is being add
+ * @param name Name of channel. Eg 'in_anglvel_x'.
+ *
+ * @return A new IIO channel handle
+ */
+struct sol_iio_channel *sol_iio_add_channel(struct sol_iio_device *device, const char *name);
+
+/**
+ * Read channel value.
+ *
+ * This method should be used if buffer is disabled,
+ * and readings will be performed on sysfs. When buffer is enabled, use
+ * sol_iio_read_buffer_channel_value() instead.
+ *
+ * @param channel IIO channel handle to be read
+ * @param value Where read value will be stored
+ *
+ * @return true if reading was performed correctly
+ */
+bool sol_iio_read_channel_value(struct sol_iio_channel *channel, double *value);
+
+/**
+ * Set callback for read operations.
+ *
+ * When buffer is enabled, reading can only
+ * be performed if there's data on buffer. A trigger is used to send values
+ * to buffer. When buffer is ready, the callback set here will be called
+ * and data can be read using sol_iio_read_buffer_channel_value().
+ *
+ * @param device IIO handle of device
+ * @param reader_cb sol_iio_reader_cb that will be called when there's data on
+ * buffer.
+ * @param data User data to be sent to reader callback.
+ */
+void sol_iio_set_reader_cb(struct sol_iio_device *device, sol_iio_reader_cb reader_cb, void *data);
```

Comment:
`const void *data`, otherwise passing literals and constants will trigger a warning.